extensions-contrib/influx-extensions/src/main/java/io/druid/data/input/influx/InfluxParser.java

```diff
@@ -0,0 +1,169 @@
+    TokenStream tokenStream = new CommonTokenStream(lexer);
+    InfluxLineProtocolParser parser = new InfluxLineProtocolParser(tokenStream);
+
+    List<InfluxLineProtocolParser.LineContext> lines = parser.lines().line();
+    if (lines.size() != 1) {
+      throw new ParseException("Multiple lines present; unable to parse more than one per record.");
+    }
+
+    Map<String, Object> out = new LinkedHashMap<>();
+
+    InfluxLineProtocolParser.LineContext line = lines.get(0);
+    String measurement = parseIdentifier(line.identifier());
+
+    if (!checkWhitelist(measurement)) {
+      throw new ParseException("Metric not whitelisted.");
+    }
+
+    out.put(MEASUREMENT_KEY, measurement);
+    if (line.tag_set() != null) {
+      line.tag_set().tag_pair().forEach(t -> parseTag(t, out));
+    }
+
+    line.field_set().field_pair().forEach(t -> parseField(t, out));
+
+    if (line.timestamp() != null) {
+      String timestamp = line.timestamp().getText();
+      parseTimestamp(timestamp, out);
+    }
+    return out;
+  }
+
+  private void parseTag(InfluxLineProtocolParser.Tag_pairContext tag, Map<String, Object> out)
+  {
+    String key = parseIdentifier(tag.identifier(0));
+    String value = parseIdentifier(tag.identifier(1));
+    out.put(key, value);
+  }
+
+  private void parseField(InfluxLineProtocolParser.Field_pairContext field, Map<String, Object> out)
+  {
+    String key = parseIdentifier(field.identifier());
+    InfluxLineProtocolParser.Field_valueContext valueContext = field.field_value();
+    Object value;
+    if (valueContext.NUMBER() != null) {
+      value = parseNumber(valueContext.NUMBER().getText());
+    } else if (valueContext.BOOLEAN() != null) {
+      value = parseBool(valueContext.BOOLEAN().getText());
+    } else {
+      value = parseQuotedString(valueContext.QUOTED_STRING().getText());
+    }
+    out.put(key, value);
+  }
+
+  private Object parseQuotedString(String text)
+  {
+    return text.substring(1, text.length() - 1).replaceAll("\\\\\"", "\"");
+  }
+
+  private Object parseNumber(String raw)
+  {
+    if (raw.endsWith("i")) {
+      return new Long(raw.substring(0, raw.length() - 1));
+    }
+
+    return new Double(raw);
+  }
+
+  // TODO: Support returning numeric value?
+  private Object parseBool(String raw)
+  {
+    char first = raw.charAt(0);
+    if (first == 't' || first == 'T') {
+      return "true";
+    } else {
+      return "false";
+    }
+  }
+
+  private String parseIdentifier(InfluxLineProtocolParser.IdentifierContext ctx)
+  {
+    if (ctx.BOOLEAN() != null || ctx.NUMBER() != null) {
+      return ctx.getText();
+    }
+
+    return ctx.IDENTIFIER_STRING().getText().replaceAll("\\\\([,= ])", "$1");
+  }
+
+  private boolean checkWhitelist(String m)
+  {
+    return (measurementWhitelist == null) || measurementWhitelist.contains(m);
+  }
+
+  private void parseTimestamp(String timestamp, Map<String, Object> dest)
+  {
+    if (timestamp.length() < 7) {
```

Comment:
Can you add a comment about the nanosecond->millisecond conversion here?

Some other comments:
- Should timestamps under 1 millisecond be treated as invalid? To me it seems like 0 would be a better return value for that case, or rounding the number to nearest milli
- The truncation could let bad timestamps slip by, e.g., if I got a timestamp "152036516185300abcd", this is probably bad data but this conversion would allow that timestamp

Reply:
Thanks for the feedback @jon-wei; I'll add comment and improve default handling there. Still waiting on the Man to sign the CLA 😁; should be pretty soon.